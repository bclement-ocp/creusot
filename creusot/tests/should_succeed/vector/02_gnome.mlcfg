
module Core_Cmp_Ordering_Type
  type t_ordering  =
    | C_Less
    | C_Equal
    | C_Greater
    
end
module CreusotContracts_Logic_Ord_OrdLogic_CmpLog_Stub
  type self
  use Core_Cmp_Ordering_Type as Core_Cmp_Ordering_Type
  function cmp_log (self : self) (_2' : self) : Core_Cmp_Ordering_Type.t_ordering
end
module CreusotContracts_Logic_Ord_OrdLogic_CmpLog_Interface
  type self
  use Core_Cmp_Ordering_Type as Core_Cmp_Ordering_Type
  function cmp_log (self : self) (_2' : self) : Core_Cmp_Ordering_Type.t_ordering
end
module CreusotContracts_Logic_Ord_OrdLogic_CmpLog
  type self
  use Core_Cmp_Ordering_Type as Core_Cmp_Ordering_Type
  function cmp_log (self : self) (_2' : self) : Core_Cmp_Ordering_Type.t_ordering
  val cmp_log (self : self) (_2' : self) : Core_Cmp_Ordering_Type.t_ordering
    ensures { result = cmp_log self _2' }
    
end
module CreusotContracts_Logic_Ord_OrdLogic_LeLog_Stub
  type self
  predicate le_log (self : self) (o : self)
end
module CreusotContracts_Logic_Ord_OrdLogic_LeLog_Interface
  type self
  predicate le_log (self : self) (o : self)
end
module CreusotContracts_Logic_Ord_OrdLogic_LeLog
  type self
  use Core_Cmp_Ordering_Type as Core_Cmp_Ordering_Type
  clone CreusotContracts_Logic_Ord_OrdLogic_CmpLog_Stub as CmpLog0 with type self = self
  predicate le_log (self : self) (o : self) =
    CmpLog0.cmp_log self o <> Core_Cmp_Ordering_Type.C_Greater
  val le_log (self : self) (o : self) : bool
    ensures { result = le_log self o }
    
end
module CreusotContracts_Logic_Seq_Impl3_Index_Stub
  type t
  use prelude.Borrow
  use seq.Seq
  use mach.int.Int
  function index (self : Seq.seq t) (_2' : int) : t
end
module CreusotContracts_Logic_Seq_Impl3_Index_Interface
  type t
  use prelude.Borrow
  use seq.Seq
  use mach.int.Int
  function index (self : Seq.seq t) (_2' : int) : t
end
module CreusotContracts_Logic_Seq_Impl3_Index
  type t
  function index (self : Seq.seq t) (_2' : Int.int) : t
  val index (self : Seq.seq t) (_2' : Int.int) : t
    ensures { result = index self _2' }
    
end
module CreusotContracts_Logic_Ord_Impl1_LtLog_Stub
  use mach.int.Int
  predicate lt_log (self : int) (_2' : int)
end
module CreusotContracts_Logic_Ord_Impl1_LtLog_Interface
  use mach.int.Int
  predicate lt_log (self : int) (_2' : int)
end
module CreusotContracts_Logic_Ord_Impl1_LtLog
  predicate lt_log (self : Int.int) (_2' : Int.int)
  val lt_log (self : Int.int) (_2' : Int.int) : bool
    ensures { result = lt_log self _2' }
    
end
module CreusotContracts_Logic_Ord_Impl1_LeLog_Stub
  use mach.int.Int
  predicate le_log (self : int) (_2' : int)
end
module CreusotContracts_Logic_Ord_Impl1_LeLog_Interface
  use mach.int.Int
  predicate le_log (self : int) (_2' : int)
end
module CreusotContracts_Logic_Ord_Impl1_LeLog
  predicate le_log (self : Int.int) (_2' : Int.int)
  val le_log (self : Int.int) (_2' : Int.int) : bool
    ensures { result = le_log self _2' }
    
end
module Core_Marker_PhantomData_Type
  type t_phantomdata 't =
    | C_PhantomData
    
end
module C02Gnome_SortedRange_Stub
  type t
  use seq.Seq
  use mach.int.Int
  predicate sorted_range (s : Seq.seq t) (l : int) (u : int)
end
module C02Gnome_SortedRange_Interface
  type t
  use seq.Seq
  use mach.int.Int
  predicate sorted_range (s : Seq.seq t) (l : int) (u : int)
end
module C02Gnome_SortedRange
  type t
  use Core_Cmp_Ordering_Type as Core_Cmp_Ordering_Type
  use CreusotContracts_Logic_Int_Int_Type as CreusotContracts_Logic_Int_Int_Type
  use Core_Marker_PhantomData_Type as Core_Marker_PhantomData_Type
  use CreusotContracts_Logic_Seq_Seq_Type as CreusotContracts_Logic_Seq_Seq_Type
  clone CreusotContracts_Logic_Seq_Impl3_Index_Stub as Index0 with type t = t
  clone CreusotContracts_Logic_Ord_OrdLogic_CmpLog_Stub as CmpLog0 with type self = t
  clone CreusotContracts_Logic_Ord_OrdLogic_LeLog_Stub as LeLog0 with type self = t, val INVALID.prior = CmpLog0.cmp_log
  clone CreusotContracts_Logic_Ord_Impl1_LtLog_Stub as LtLog0
  clone CreusotContracts_Logic_Ord_Impl1_LeLog_Stub as LeLog0
  predicate sorted_range [#"../02_gnome.rs" 6 0 6 58] (s : Seq.seq t) (l : Int.int) (u : Int.int) =
    [#"../02_gnome.rs" 7 4 9 5] forall j : Int.int . forall i : Int.int . l <= i /\ i < j /\ j < u -> LeLog0.le_log (Seq.get s i) (Seq.get s j)
  val sorted_range [#"../02_gnome.rs" 6 0 6 58] (s : Seq.seq t) (l : Int.int) (u : Int.int) : bool
    ensures { result = sorted_range s l u }
    
end
module CreusotContracts_Logic_Seq_Impl2_Len_Stub
  type t
  use seq.Seq
  use mach.int.Int
  function len (self : Seq.seq t) : int
end
module CreusotContracts_Logic_Seq_Impl2_Len_Interface
  type t
  use seq.Seq
  use mach.int.Int
  function len (self : Seq.seq t) : int
end
module CreusotContracts_Logic_Seq_Impl2_Len
  type t
  function len (self : Seq.seq t) : Int.int
  val len (self : Seq.seq t) : Int.int
    ensures { result = len self }
    
end
module CreusotContracts_Logic_Int_Impl27_From_Stub
  use mach.int.Int
  use mach.int.Int32
  function from (_1' : int32) : int
end
module CreusotContracts_Logic_Int_Impl27_From_Interface
  use mach.int.Int
  use mach.int.Int32
  function from (_1' : int32) : int
end
module CreusotContracts_Logic_Int_Impl27_From
  function from (_1' : int32) : Int.int
  val from (_1' : int32) : Int.int
    ensures { result = from _1' }
    
end
module C02Gnome_Sorted_Stub
  type t
  use seq.Seq
  predicate sorted (s : Seq.seq t)
end
module C02Gnome_Sorted_Interface
  type t
  use seq.Seq
  predicate sorted (s : Seq.seq t)
end
module C02Gnome_Sorted
  type t
  use Core_Cmp_Ordering_Type as Core_Cmp_Ordering_Type
  use CreusotContracts_Logic_Int_Int_Type as CreusotContracts_Logic_Int_Int_Type
  use Core_Marker_PhantomData_Type as Core_Marker_PhantomData_Type
  use CreusotContracts_Logic_Seq_Seq_Type as CreusotContracts_Logic_Seq_Seq_Type
  clone CreusotContracts_Logic_Seq_Impl2_Len_Stub as Len0 with type t = t
  clone CreusotContracts_Logic_Int_Impl27_From_Stub as From0
  clone CreusotContracts_Logic_Seq_Impl3_Index_Stub as Index0 with type t = t
  clone CreusotContracts_Logic_Ord_OrdLogic_CmpLog_Stub as CmpLog0 with type self = t
  clone CreusotContracts_Logic_Ord_OrdLogic_LeLog_Stub as LeLog0 with type self = t, val INVALID.prior = CmpLog0.cmp_log
  clone CreusotContracts_Logic_Ord_Impl1_LtLog_Stub as LtLog0
  clone CreusotContracts_Logic_Ord_Impl1_LeLog_Stub as LeLog0
  clone C02Gnome_SortedRange_Stub as SortedRange0 with type t = t, val INVALID.prior = LeLog0.le_log,
    val INVALID.prior = LtLog0.lt_log, val INVALID.prior = LeLog0.le_log, val INVALID.prior = Index0.index
  predicate sorted [#"../02_gnome.rs" 13 0 13 36] (s : Seq.seq t) =
    [#"../02_gnome.rs" 14 4 14 31] SortedRange0.sorted_range s 0 (Seq.length s)
  val sorted [#"../02_gnome.rs" 13 0 13 36] (s : Seq.seq t) : bool
    ensures { result = sorted s }
    
end
