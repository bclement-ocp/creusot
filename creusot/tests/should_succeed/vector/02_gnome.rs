extern crate creusot_contracts;

use creusot_contracts::{std::cmp::Ord, *};

#[predicate]
fn sorted_range<T: Ord>(s: Seq<T>, l: Int, u: Int) -> bool {
    pearlite! {
        forall<i : Int, j : Int> l <= i && i < j && j < u ==> s[i] <= s[j]
    }
}

#[predicate]
fn sorted<T: Ord>(s: Seq<T>) -> bool {
    sorted_range(s, 0, s.len())
}

// #[ensures(sorted(@^v))]
// #[ensures((@^v).permutation_of(@*v))]
// pub fn gnome_sort<T: Ord>(v: &mut Vec<T>) {
//     let old_v = ghost! { v };
//     let mut i = 0;
//     #[invariant(sorted, sorted_range(@v, 0, @i))]
//     #[invariant(proph_const, ^v == ^old_v.inner())]
//     #[invariant(permutation, (@*v).permutation_of(@*old_v.inner()))]
//     while i < v.len() {
//         if i == 0 || v[i - 1].le(&v[i]) {
//             i += 1;
//         } else {
//             v.swap(i - 1, i);
//             i -= 1;
//         }
//     }
// }
