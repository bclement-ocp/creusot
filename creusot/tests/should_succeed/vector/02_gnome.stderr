---- DefId(20:679 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::cmp_log) ----
--------
---- DefId(20:680 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::le_log) ----
--------
---- DefId(0:9 ~ 02_gnome[1736]::sorted_range) ----
---- DefId(20:858 ~ creusot_contracts[be3c]::logic::seq::{impl#3}::index) ----
--------
---- DefId(20:709 ~ creusot_contracts[be3c]::logic::ord::{impl#1}::lt_log) ----
--------
---- DefId(20:708 ~ creusot_contracts[be3c]::logic::ord::{impl#1}::le_log) ----
--------
--------
---- DefId(0:13 ~ 02_gnome[1736]::sorted) ----
---- DefId(20:847 ~ creusot_contracts[be3c]::logic::seq::{impl#2}::len) ----
--------
---- DefId(20:639 ~ creusot_contracts[be3c]::logic::int::{impl#27}::from) ----
--------
--------
---- DefId(20:2240 ~ creusot_contracts[be3c]::std#1::vec::{impl#0}::model) ----
--------
---- DefId(20:852 ~ creusot_contracts[be3c]::logic::seq::{impl#2}::permutation_of) ----
--------
---- DefId(20:2140 ~ creusot_contracts[be3c]::std#1::slice::SliceIndexSpec::in_bounds) ----
--------
---- DefId(20:2141 ~ creusot_contracts[be3c]::std#1::slice::SliceIndexSpec::has_value) ----
--------
---- DefId(20:2130 ~ creusot_contracts[be3c]::std#1::slice::{impl#0}::model) ----
--------
---- DefId(20:671 ~ creusot_contracts[be3c]::logic::model::Model::model) ----
--------
---- DefId(20:673 ~ creusot_contracts[be3c]::logic::model::{impl#1}::model) ----
--------
---- DefId(20:826 ~ creusot_contracts[be3c]::logic::resolve::Resolve::resolve) ----
--------
---- DefId(20:582 ~ creusot_contracts[be3c]::logic::ghost::{impl#1}::inner) ----
--------
---- DefId(20:833 ~ creusot_contracts[be3c]::logic::resolve::{impl#1}::resolve) ----
--------
---- DefId(20:2143 ~ creusot_contracts[be3c]::std#1::slice::{impl#3}::in_bounds) ----
--------
---- DefId(20:2144 ~ creusot_contracts[be3c]::std#1::slice::{impl#3}::has_value) ----
--------
---- DefId(20:682 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::cmp_le_log) ----
--------
---- DefId(20:683 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::lt_log) ----
--------
---- DefId(20:685 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::cmp_lt_log) ----
--------
---- DefId(20:686 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::ge_log) ----
--------
---- DefId(20:688 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::cmp_ge_log) ----
--------
---- DefId(20:689 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::gt_log) ----
--------
---- DefId(20:691 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::cmp_gt_log) ----
--------
---- DefId(20:693 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::refl) ----
--------
---- DefId(20:697 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::trans) ----
--------
---- DefId(20:700 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::antisym1) ----
--------
---- DefId(20:703 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::antisym2) ----
--------
---- DefId(20:705 ~ creusot_contracts[be3c]::logic::ord::OrdLogic::eq_cmp) ----
--------
